--- a/elan3/src/elan-compiler/src/rem/parser/REFParser.java
+++ b/elan3/src/elan-compiler/src/rem/parser/REFParser.java
@@ -1677,8 +1677,8 @@
         jj_expentry[i] = jj_lasttokens[i];
       }
       boolean exists = false;
-      for (java.util.Enumeration enum = jj_expentries.elements(); enum.hasMoreElements();) {
-        int[] oldentry = (int[])(enum.nextElement());
+      for (java.util.Enumeration jj_enum = jj_expentries.elements(); jj_enum.hasMoreElements();) {
+        int[] oldentry = (int[])(jj_enum.nextElement());
         if (oldentry.length == jj_expentry.length) {
           exists = true;
           for (int i = 0; i < jj_expentry.length; i++) {
